File: LeetCode - Algorithm/Matrix/73.js
// 73. Set Matrix Zeroes
// https://leetcode.com/problems/set-matrix-zeroes/

/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function (matrix) {
  const numRows = matrix.length;
  const numCols = matrix[0].length;

  const initialZeros = [];

  for (let i = 0; i < numRows; i++) {
    for (let j = 0; j < numCols; j++) {
      if (matrix[i][j] === 0) initialZeros.push([i, j]);
    }
  }

  initialZeros.forEach(([xCoordinate, yCoordinate]) => {
    for (let i = 0; i < numRows; i++) {
      matrix[i][yCoordinate] = 0;
    }

    for (let i = 0; i < numCols; i++) {
      matrix[xCoordinate][i] = 0;
    }
  });
};

File: LeetCode - Algorithm/Matrix/73.java
// 73. Set Matrix Zeroes
// https://leetcode.com/problems/set-matrix-zeroes/

import java.util.Set;
import java.util.HashSet;

class Solution {
    public void setZeroes(int[][] matrix) {
        Set<Integer> rows = new HashSet<>();
        Set<Integer> cols = new HashSet<>();

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                if (matrix[i][j] == 0) {
                    rows.add(i);
                    cols.add(j);
                }
            }
        }

        for (int row : rows)
            for (int j = 0; j < matrix[0].length; j++)
                matrix[row][j] = 0;

        for (int col : cols)
            for (int i = 0; i < matrix.length; i++)
                matrix[i][col] = 0;
    }
}
